### PR TITLE
Suggest to swap a struct and a trait in trait impls

### LINKED ISSUE #89590: Suggest to swap a struct and a trait in trait impls where appropriate
Given the following code: [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d1ef78517b400c10a8fc7ed8149ce8c4)

```rust
pub trait Trait {}

pub struct Struct {}

impl Struct for Trait {}

```

The current output is:

```
error[E0404]: expected trait, found struct `Struct`
 --> src/lib.rs:5:6
  |
5 | impl Struct for Trait {}
  |      ^^^^^^ not a trait

warning: trait objects without an explicit `dyn` are deprecated
 --> src/lib.rs:5:17
  |
5 | impl Struct for Trait {}
  |                 ^^^^^ help: use `dyn`: `dyn Trait`
  |
  = note: `#[warn(bare_trait_objects)]` on by default
  = warning: this is accepted in the current edition (Rust 2018) but is a hard error in Rust 2021!
  = note: for more information, see issue #80165 <https://github.com/rust-lang/rust/issues/80165>
```

### PR DESCRIPTION
**Pull request by TaKO8Ki**
closes #89590

### CODE REVIEW

**Comment by TaKO8Ki** [comment]
@petrochenkov 
Thank you for your review. I would like to know what others think about this issue. @rust-lang/wg-diagnostics

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 1]
```diff
@@ -248,6 +250,25 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+            && trait_ref.path.span == span
+            && let PathSource::Trait(_) = source
+            && let Some(Res::Def(DefKind::Struct, _)) = res
+            && let Ok(self_ty_str) =
+                self.r.session.source_map().span_to_snippet(self_ty.span)
+            && let Ok(trait_ref_str) =
+                self.r.session.source_map().span_to_snippet(trait_ref.path.span)
+        {
+                err.multipart_suggestion(
+                    "consider swapping the struct and the trait",
```
Comment (estebank) [inline compiler/rustc_resolve/src/late/diagnostics.rs:267]: If we do land the suggestion, I would use more "forceful" wording here
```suggestion
                    "`impl` items mention the trait being implemented first and the type it is being implemented for second",
```

File: src/test/ui/suggestions/suggest-swapping-self-ty-and-trait-edition-2021.stderr [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+error[E0782]: trait objects must include the `dyn` keyword
+  --> $DIR/suggest-swapping-self-ty-and-trait-edition-2021.rs:7:27
+   |
+LL | impl<'a, T> Struct<T> for Trait<'a, T> {}
+   |                           ^^^^^^^^^^^^
+   |
+help: add `dyn` keyword before this trait
+   |
+LL - impl<'a, T> Struct<T> for Trait<'a, T> {}
+LL + impl<'a, T> Struct<T> for dyn Trait<'a, T> {}
```
Comment (estebank) [inline src/test/ui/suggestions/suggest-swapping-self-ty-and-trait-edition-2021.stderr:21]: Ideally we'd silence this follow up error

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 3]
```diff
@@ -248,6 +250,25 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+            && let TyKind::Path(_, self_ty_path) = &self_ty.kind
+            && let PathResult::Module(ModuleOrUniformRoot::Module(module)) = self.resolve_path(&Segment::from_path(self_ty_path), Some(TypeNS), None)
+            && let ModuleKind::Def(DefKind::Trait, ..) = module.kind
+            && trait_ref.path.span == span
+            && let PathSource::Trait(_) = source
+            && let Some(Res::Def(DefKind::Struct, _)) = res
+            && let Ok(self_ty_str) =
+                self.r.session.source_map().span_to_snippet(self_ty.span)
+            && let Ok(trait_ref_str) =
+                self.r.session.source_map().span_to_snippet(trait_ref.path.span)
```
Comment (estebank) [inline compiler/rustc_resolve/src/late/diagnostics.rs:264]: This is quite a bit `let` expr chain. I would at least add a comment for it, if not move it to its own method with a descriptive name.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 4]
```diff
@@ -248,6 +250,25 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
         let mut err =
             self.r.session.struct_span_err_with_code(base_error.span, &base_error.msg, code);
+        if let Some((trait_ref, self_ty)) =
+            self.diagnostic_metadata.currently_processing_impl_trait.clone()
+            && let TyKind::Path(_, self_ty_path) = &self_ty.kind
+            && let PathResult::Module(ModuleOrUniformRoot::Module(module)) = self.resolve_path(&Segment::from_path(self_ty_path), Some(TypeNS), None)
+            && let ModuleKind::Def(DefKind::Trait, ..) = module.kind
+            && trait_ref.path.span == span
+            && let PathSource::Trait(_) = source
+            && let Some(Res::Def(DefKind::Struct, _)) = res
```
Comment (estebank) [inline compiler/rustc_resolve/src/late/diagnostics.rs:260]: You wouldn't want to restrict this to structs only, but support enums, unions and primitive types as well.

**Review by estebank — COMMENTED** [review]
If we could either move the logic out of the way so that the code is easier to read, or simplify and only signal that we're resolving an `impl` item header and emit not a structured suggestion but a note, it would be better.

@petrochenkov are you concerned with the growth of `DiagnosticMetadata`, the added logic, or both?

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 3 above)
Comment (TaKO8Ki) [reply compiler/rustc_resolve/src/late/diagnostics.rs]: I fixed that.

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 4 above)
Comment (TaKO8Ki) [reply compiler/rustc_resolve/src/late/diagnostics.rs]: I fixed that.

File: src/test/ui/suggestions/suggest-swapping-self-ty-and-trait-edition-2021.stderr (hunk 2 above)
Comment (TaKO8Ki) [reply src/test/ui/suggestions/suggest-swapping-self-ty-and-trait-edition-2021.stderr]: Can I open a new issue about that and work on it later?

Comment (estebank) [reply src/test/ui/suggestions/suggest-swapping-self-ty-and-trait-edition-2021.stderr]: @TaKO8Ki absolutely!

Comment (TaKO8Ki) [reply src/test/ui/suggestions/suggest-swapping-self-ty-and-trait-edition-2021.stderr]: I opened a issue. #97985

**Pull request merged**